LE Audio: BAP long write

Thalley opened the pull request:
Adds support for long writes as the BAP Unicast Client.

Similar additions may be needed for the Broadcast Assistant 

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/54367
depends on https://github.com/zephyrproject-rtos/zephyr/pull/56108
CI issue depends on https://github.com/zephyrproject-rtos/zephyr/pull/58180

Thalley committed:
Bluetooth: BAP: Rename BAP uni cli read buf to att_buf

att_buf is more generic and makes more sense when we
also want to use it for long writes.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: BAP: Add unicast client and server write long support

Add support for long writes for the unicast client and server.
This reuses the ATT buffer for long reads.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Tests: Bluetooth: Audio: Add write of long metadata

Add test of long metadata write.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: BAP: Add support for long ASE notifications

Add support for long notifications ASE notifications
by doing a long read if the notification has length of
the maximum MTU.

This behavior is currently not defined by the BAP spec,
so may not work with all devices.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

larsgk reviewed (dismissed):
I would like to try it out but LGTM

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -27,6 +27,8 @@ LOG_MODULE_REGISTER(bt_ascs, CONFIG_BT_ASCS_LOG_LEVEL);
 #include "common/bt_str.h"
 #include "common/assert.h"
 
+#include "../host/att_internal.h"
+
```
I agree, but this will require significant changes to ATT. I'll rather do that in a separate PR. If that PR merges before this, I'll update this PR; vice versa if this merges first I'll modify this in that PR.

Thalley committed:
Bluetooth: BAP: Unicast client optimize long notify reads

Optimize the read procedure when receiving long notifications
by storing the already notified data in the buffer if the
client instance is not already busy. If it is busy we have
no way to store it, and will just need to read the entire thing
later.

This also modifies when we reset the ATT buffer. Instead of reseting
when we want to use it, we reset it when we are done using it.
The reason for this is to avoid making state checks before calling
long_ase_read to determine whether it should reset the buffer
or not.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/audio/bap_unicast_client.c:
```diff
@@ -1814,14 +1953,52 @@ static int unicast_client_ep_release(struct bt_bap_ep *ep, struct net_buf_simple
 
 	LOG_DBG("conn %p ep %p buf %p len %u", conn, ep, buf, buf->len);
 
+	if (buf->len > max_write_size) {
```
Yes! Thanks.

Thalley commented on subsys/bluetooth/audio/bap_unicast_client.c:
```diff
@@ -1398,23 +1406,151 @@ static uint8_t unicast_client_cp_notify(struct bt_conn *conn,
 		return BT_GATT_ITER_STOP;
 	}
 
+	if (length == max_ntf_size) {
+		long_ase_read(client_ep);
+
+		return BT_GATT_ITER_CONTINUE;
+	}
```
Agreed. I added a new commit to change it to do this, as it required quite a few of refactoring to get to work. This also uncovered a few bugs which are fixed by the same commit

Thalley commented:
Added mandatory ATT_PREPARE_COUNT to the recently merged TMAP peripheral
